core/src/main/java/org/apache/calcite/avatica/AvaticaUtils.java

```diff
@@ -222,15 +223,22 @@ public static <T> T instantiatePlugin(Class<T> pluginClass,
         final Class<T> clazz = (Class) Class.forName(left);
         final Field field;
         field = clazz.getField(right);
-        return pluginClass.cast(field.get(null));
+        final Object fieldValue = field.get(null);
+        if (fieldValue instanceof ThreadLocal) {
+          value = ((ThreadLocal<?>) fieldValue).get();
+        } else {
+          value = fieldValue;
+        }
+        return pluginClass.cast(value);
       }
       //noinspection unchecked
       final Class<T> clazz = (Class) Class.forName(className);
       try {
         // We assume that if there is an INSTANCE field it is static and
         // has the right type.
         final Field field = clazz.getField("INSTANCE");
-        return pluginClass.cast(field.get(null));
+        value = field.get(null);
+        return pluginClass.cast(value);
```

Comment:
In this branch, `value` could also be a `ThreadLocal`?

Reply:
No. If a class `Foo` has a field called `INSTANCE`, it should be of type `Foo`. It would be confusing for a field called `INSTANCE` to be of any other type.

Reply:
Good question, though.

Reply:
Agreed, thanks for the explanation.